File: Minimal/src/Brontide.Minimal.Host/ProviderServingTrustRevalidation.fs
namespace Brontide.Minimal.Host

open System
open Brontide.Minimal.Binding.Portable
open Brontide.Minimal.Experimental.ComponentManagement

type ProviderServingTrustRevalidationResult =
    { Code: string
      RefusedBy: string
      Revalidated: bool
      Continued: bool
      LaunchPolicyIdentity: ProviderPublisherTrustPolicyId option
      ServingPolicyIdentity: ProviderPublisherTrustPolicyId option
      Authorization: TrustedProviderPublisherAuthorization option
      RetirementCode: string }

type ProviderServingActivation =
    private
        { Chain: ProviderDistributionChainResult
          Lifecycle: ComponentBindingLifecycleResult option
          Occurrence: OccurrenceId
          Resolution: ResolutionOutcome
          Selection: ComponentBindingSelection
          Request: ActivationRuntimeRequest
          mutable RestartState: int }
    member this.OccurrenceId = this.Occurrence
    member internal this.DistributionChain = this.Chain
    member internal this.BindingLifecycle = this.Lifecycle
    member internal this.RetainedResolution = this.Resolution
    member internal this.RetainedSelection = this.Selection
    member internal this.RetainedRequest = this.Request
    member internal this.BeginRestart() =
        lock this (fun () ->
            match this.RestartState with
            | 0 -> this.RestartState <- 1; "restart-claimed"
            | 1 -> "provider-restart-in-progress"
            | _ -> "provider-restart-already-completed")
    member internal this.FinishRestart completed =
        lock this (fun () -> this.RestartState <- if completed then 2 else 0)
    member internal this.Restarted(chain, lifecycle) =
        { Chain = chain
          Lifecycle = Some lifecycle
          Occurrence = this.Occurrence
          Resolution = this.Resolution
          Selection = this.Selection
          Request = this.Request
          RestartState = 0 }
    member this.IsServing =
        this.Chain.Provider |> Option.exists (fun provider -> not provider.HasExited)
        && this.Lifecycle |> Option.exists (fun lifecycle ->
            lifecycle.Failure.IsNone
            && lifecycle.Member |> Option.exists _.IsReleased
            && lifecycle.Runtime |> Option.exists (fun runtime -> runtime.Kind = ActivationRuntimeOutcomeKind.Active))
    member this.MemberReleased =
        this.Lifecycle |> Option.bind _.Member |> Option.exists _.IsReleased
    member this.Retire reason =
        task {
            if String.IsNullOrWhiteSpace reason then invalidArg (nameof reason) "retirement reason is required"
            match this.Lifecycle |> Option.bind _.Member with
            | Some memberValue when memberValue.IsReleased ->
                let! _ = memberValue.Retire reason
                return ()
            | _ -> return ()
        }

/// Takes one current publisher-trust decision for an already released portable member. A lapsed
/// publisher is retired and its concrete provider is terminated; cadence remains a host concern.
[<RequireQualifiedAccess>]
module ProviderServingTrustRevalidation =
    let activate
        (chain: ProviderDistributionChainResult)
        resolution
        selection
        request
        =
        task {
            match chain.Provider with
            | Some provider when chain.Revalidated ->
                let! lifecycle =
                    ComponentBindingLifecycle.activate resolution selection request provider.Conversation
                return
                    { Chain = chain; Lifecycle = Some lifecycle; Occurrence = selection.Occurrence
                      Resolution = resolution; Selection = selection; Request = request; RestartState = 0 }
            | _ ->
                return
                    { Chain = chain; Lifecycle = None; Occurrence = selection.Occurrence
                      Resolution = resolution; Selection = selection; Request = request; RestartState = 0 }
        }

    let private revalidateCore
        (registry: DurableProviderPublisherTrustPolicyRegistry)
        (store: ContentAddressedProviderStore)
        (activation: ProviderServingActivation)
        retirementReason
        removeStagedSet
        =
        task {
            if isNull (box registry) then nullArg (nameof registry)
            if isNull (box store) then nullArg (nameof store)
            if String.IsNullOrWhiteSpace retirementReason then
                invalidArg (nameof retirementReason) "retirement reason is required"

            let chain = activation.Chain

            match
                chain.Provider,
                chain.VerifiedEvidence,
                chain.StagedIdentity,
                chain.PolicyAuthorityIdentity,
                activation.Lifecycle |> Option.bind _.Member
            with
            | Some provider, Some evidence, Some stagedIdentity, Some authority, Some memberValue
                when chain.Revalidated
                     && authority = registry.AuthorityIdentity
                     && not provider.HasExited
                     && memberValue.IsReleased
                     && activation.IsServing ->
                match registry.Current with
                | None ->
                    return
                        { Code = "publisher-trust-policy-unavailable"; RefusedBy = "cbi37"
                          Revalidated = false; Continued = false
                          LaunchPolicyIdentity = chain.LaunchPolicyIdentity
                          ServingPolicyIdentity = None; Authorization = None
                          RetirementCode = "retirement-not-attempted" }
                | Some current ->
                    let trust = ProviderPublisherTrustEvaluator.evaluate current.Policy (Some evidence)
                    match trust.Authorization with
                    | Some authorization ->
                        return
                            { Code = "publisher-trust-current"; RefusedBy = "none"
                              Revalidated = true; Continued = true
                              LaunchPolicyIdentity = chain.LaunchPolicyIdentity
                              ServingPolicyIdentity = Some current.Policy.Identity
                              Authorization = Some authorization
                              RetirementCode = "retirement-not-attempted" }
                    | None ->
                        let! retired = memberValue.Retire retirementReason
                        let mutable retirementCode =
                            match retired with Ok _ -> "retired" | Error _ -> "retirement-failed"
                        provider.Dispose()
                        if removeStagedSet then
                            let removal = store.Remove stagedIdentity
                            if not removal.Removed && removal.Code <> "artifact-set-not-staged" then
                                retirementCode <- $"{retirementCode};{removal.Code}"
                        return
                            { Code = trust.Code; RefusedBy = "cbi35"
                              Revalidated = true; Continued = false
                              LaunchPolicyIdentity = chain.LaunchPolicyIdentity
                              ServingPolicyIdentity = Some current.Policy.Identity
                              Authorization = None; RetirementCode = retirementCode }
            | _ ->
                return
                    { Code = "serving-activation-unavailable"; RefusedBy = "none"
                      Revalidated = false; Continued = false
                      LaunchPolicyIdentity = chain.LaunchPolicyIdentity
                      ServingPolicyIdentity = None; Authorization = None
                      RetirementCode = "retirement-not-attempted" }
        }

    let revalidate registry store activation retirementReason =
        revalidateCore registry store activation retirementReason true

    let internal revalidateForSweep registry store activation retirementReason =
        revalidateCore registry store activation retirementReason false

type ProviderServingTrustSweepMember =
    { Occurrence: OccurrenceId
      Result: ProviderServingTrustRevalidationResult }

type ProviderServingTrustSweepResult =
    { Code: string
      RefusedBy: string
      Members: ProviderServingTrustSweepMember list
      ContinuedCount: int
      WithdrawnCount: int }

/// Applies one deterministic, bounded host-owned trust sweep. Invocation cadence remains external.
[<RequireQualifiedAccess>]
module ProviderServingTrustSweep =
    [<Literal>]
    let MaximumMembers = 64

    let runRecording
        (registry: DurableProviderPublisherTrustPolicyRegistry)
        (store: ContentAddressedProviderStore)
        (activations: ProviderServingActivation list)
        retirementReason
        (attributions: DurableProviderStopAttributionStore option)
        (recordedAt: DateTimeOffset)
        =
        task {
            if String.IsNullOrWhiteSpace retirementReason then
                invalidArg (nameof retirementReason) "retirement reason is required"

            let distinct =
                activations
                |> List.map _.OccurrenceId
                |> List.distinct
                |> List.length

            if List.isEmpty activations
               || List.length activations > MaximumMembers
               || activations |> List.exists (fun activation -> not activation.IsServing)
               || distinct <> List.length activations then
                return
                    { Code = "serving-trust-sweep-invalid"; RefusedBy = "preflight"
                      Members = []; ContinuedCount = 0; WithdrawnCount = 0 }
            else
                if isNull (box registry) then nullArg (nameof registry)
                if isNull (box store) then nullArg (nameof store)

                let ordered =
                    activations
                    |> List.sortWith (fun left right ->
                        String.CompareOrdinal(
                            OccurrenceId.value left.OccurrenceId,
                            OccurrenceId.value right.OccurrenceId))
                let members = ResizeArray<ProviderServingTrustSweepMember>()
                for activation in ordered do
                    let! result =
                        ProviderServingTrustRevalidation.revalidateForSweep registry store activation retirementReason
                    members.Add { Occurrence = activation.OccurrenceId; Result = result }
                    // After the effect, never before: the record states that this member was stopped.
                    match attributions, activation.DistributionChain.StagedIdentity with
                    | Some store, Some staged when not result.Continued ->
                        store.Record(activation.OccurrenceId, staged, recordedAt, PublisherTrustWithdrawal)
                        |> ignore
                    | _ -> ()
                let grouped =
                    ordered
                    |> List.choose (fun activation ->
                        activation.Chain.StagedIdentity
                        |> Option.map (fun staged -> staged, activation.OccurrenceId))
                    |> List.groupBy fst
                for stagedIdentity, groupedMembers in grouped do
                    let occurrences = groupedMembers |> List.map snd |> Set.ofList
                    let mustRetain =
                        members
                        |> Seq.exists (fun memberValue ->
                            Set.contains memberValue.Occurrence occurrences
                            && (memberValue.Result.Continued || not memberValue.Result.Revalidated))
                    if not mustRetain then
                        let removal = store.Remove stagedIdentity
                        if not removal.Removed && removal.Code <> "artifact-set-not-staged" then
                            for index in 0 .. members.Count - 1 do
                                let memberValue = members[index]
                                if Set.contains memberValue.Occurrence occurrences
                                   && not memberValue.Result.Continued then
                                    members[index] <-
                                        { memberValue with
                                            Result =
                                                { memberValue.Result with
                                                    RetirementCode =
                                                        $"{memberValue.Result.RetirementCode};{removal.Code}" } }
                let observations = List.ofSeq members
                let continued = observations |> List.sumBy (fun memberValue -> if memberValue.Result.Continued then 1 else 0)
                let code =
                    if observations |> List.exists (fun memberValue -> not memberValue.Result.Revalidated) then
                        "serving-trust-sweep-incomplete"
                    elif observations |> List.exists (fun memberValue ->
                        not memberValue.Result.Continued && memberValue.Result.RetirementCode <> "retired") then
                        "serving-trust-sweep-cleanup-incomplete"
                    elif continued = observations.Length then
                        "serving-trust-sweep-current"
                    else
                        "serving-trust-sweep-withdrawn"
                return
                    { Code = code; RefusedBy = "none"; Members = observations
                      ContinuedCount = continued; WithdrawnCount = observations.Length - continued }
        }

    let run registry store activations retirementReason =
        runRecording registry store activations retirementReason None DateTimeOffset.UnixEpoch

File: Reference/src/Brontide.Reference.Studio/ProviderServingTrustRevalidation.cs
namespace Brontide.Reference.Studio;

public sealed record ProviderServingTrustRevalidationResult(
    string Code,
    string RefusedBy,
    bool Revalidated,
    bool Continued,
    ProviderPublisherTrustPolicyId? LaunchPolicyIdentity,
    ProviderPublisherTrustPolicyId? ServingPolicyIdentity,
    TrustedProviderPublisherAuthorization? Authorization,
    string RetirementCode);

public sealed class ProviderServingActivation : IAsyncDisposable
{
    private readonly object _restartSync = new();
    private int _restartState;

    internal ProviderServingActivation(
        ProviderDistributionChainResult chain,
        ComponentBindingLifecycleResult? lifecycle,
        Brontide.Reference.Experimental.ComponentManagement.OccurrenceId occurrence,
        Brontide.Reference.Experimental.ComponentManagement.ResolutionOutcome resolution,
        ComponentBindingSelection selection,
        Brontide.Reference.Experimental.ComponentManagement.ActivationRuntimeRequest request)
    {
        Chain = chain;
        Lifecycle = lifecycle;
        Occurrence = occurrence;
        Resolution = resolution;
        Selection = selection;
        Request = request;
    }

    internal ProviderDistributionChainResult Chain { get; }
    internal ComponentBindingLifecycleResult? Lifecycle { get; }
    internal Brontide.Reference.Experimental.ComponentManagement.ResolutionOutcome Resolution { get; }
    internal ComponentBindingSelection Selection { get; }
    internal Brontide.Reference.Experimental.ComponentManagement.ActivationRuntimeRequest Request { get; }

    public Brontide.Reference.Experimental.ComponentManagement.OccurrenceId Occurrence { get; }

    public bool IsServing =>
        Chain.Provider is { HasExited: false } && Lifecycle?.IsActive == true;

    public bool MemberReleased => Lifecycle?.Member?.IsReleased == true;

    internal string BeginRestart()
    {
        lock (_restartSync)
        {
            return _restartState switch
            {
                0 => Claim(),
                1 => "provider-restart-in-progress",
                _ => "provider-restart-already-completed",
            };
        }

        string Claim()
        {
            _restartState = 1;
            return "restart-claimed";
        }
    }

    internal void FinishRestart(bool completed)
    {
        lock (_restartSync)
        {
            _restartState = completed ? 2 : 0;
        }
    }

    public async ValueTask RetireAsync(string reason, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(reason);
        if (Lifecycle?.Member is { IsReleased: true } member)
        {
            await member.RetireAsync(reason, cancellationToken).ConfigureAwait(false);
        }
    }

    public async ValueTask DisposeAsync()
    {
        if (Lifecycle?.Member is { } member)
        {
            await member.DisposeAsync().ConfigureAwait(false);
        }
    }
}

/// <summary>
/// Takes one current publisher-trust decision for an already released portable member. A lapsed
/// publisher is retired and its concrete provider is terminated; cadence remains a host concern.
/// </summary>
public static class ProviderServingTrustRevalidation
{
    public static async ValueTask<ProviderServingActivation> ActivateAsync(
        ProviderDistributionChainResult chain,
        Brontide.Reference.Experimental.ComponentManagement.ResolutionOutcome resolution,
        ComponentBindingSelection selection,
        Brontide.Reference.Experimental.ComponentManagement.ActivationRuntimeRequest request,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(chain);
        if (chain.Provider is not { } provider || !chain.Revalidated)
        {
            return new(chain, null, selection.Occurrence, resolution, selection, request);
        }

        var lifecycle = await ComponentBindingLifecycle.ActivateAsync(
            resolution, selection, request, provider.Conversation, cancellationToken).ConfigureAwait(false);
        return new(chain, lifecycle, selection.Occurrence, resolution, selection, request);
    }

    public static async ValueTask<ProviderServingTrustRevalidationResult> RevalidateAsync(
        DurableProviderPublisherTrustPolicyRegistry registry,
        ContentAddressedProviderStore store,
        ProviderServingActivation activation,
        string retirementReason,
        CancellationToken cancellationToken = default) =>
        await RevalidateAsync(
            registry, store, activation, retirementReason, removeStagedSet: true, cancellationToken)
            .ConfigureAwait(false);

    internal static async ValueTask<ProviderServingTrustRevalidationResult> RevalidateAsync(
        DurableProviderPublisherTrustPolicyRegistry registry,
        ContentAddressedProviderStore store,
        ProviderServingActivation activation,
        string retirementReason,
        bool removeStagedSet,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(registry);
        ArgumentNullException.ThrowIfNull(store);
        ArgumentNullException.ThrowIfNull(activation);
        ArgumentException.ThrowIfNullOrWhiteSpace(retirementReason);

        var chain = activation.Chain;
        var lifecycle = activation.Lifecycle;

        if (!chain.IsLaunched || !chain.Revalidated || chain.Provider is not { } provider
            || chain.VerifiedEvidence is not { } evidence || chain.StagedIdentity is null
            || chain.PolicyAuthorityIdentity != registry.AuthorityIdentity || provider.HasExited
            || lifecycle?.IsActive != true || lifecycle.Member is not { IsReleased: true } member)
        {
            return new(
                "serving-activation-unavailable", "none", false, false,
                chain.LaunchPolicyIdentity, null, null, "retirement-not-attempted");
        }

        var current = registry.Current;
        if (current is null)
        {
            // A matching registry that launched this provider cannot clear its current snapshot.
            // Keep the boundary fail closed if an implementation later gains such a transition.
            return new(
                "publisher-trust-policy-unavailable", "cbi37", false, false,
                chain.LaunchPolicyIdentity, null, null, "retirement-not-attempted");
        }

        var trust = ProviderPublisherTrustEvaluator.Evaluate(current.Policy, evidence);
        if (trust.Authorization is not null)
        {
            return new(
                "publisher-trust-current", "none", true, true,
                chain.LaunchPolicyIdentity, current.Policy.Identity, trust.Authorization,
                "retirement-not-attempted");
        }

        var retirementCode = "retired";
        try
        {
            await member.RetireAsync(retirementReason, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception)
        {
            // Trust withdrawal still terminates the concrete provider. The separate cleanup code
            // preserves the failed graceful-retirement observation without reopening service.
            retirementCode = "retirement-failed";
        }

        await provider.DisposeAsync().ConfigureAwait(false);
        if (removeStagedSet)
        {
            var removal = store.Remove(chain.StagedIdentity.Value);
            if (!removal.Removed && removal.Code != "artifact-set-not-staged")
            {
                retirementCode = $"{retirementCode};{removal.Code}";
            }
        }

        return new(
            trust.Code, "cbi35", true, false,
            chain.LaunchPolicyIdentity, current.Policy.Identity, null, retirementCode);
    }
}

public sealed record ProviderServingTrustSweepMember(
    Brontide.Reference.Experimental.ComponentManagement.OccurrenceId Occurrence,
    ProviderServingTrustRevalidationResult Result);

public sealed record ProviderServingTrustSweepResult(
    string Code,
    string RefusedBy,
    IReadOnlyList<ProviderServingTrustSweepMember> Members,
    int ContinuedCount,
    int WithdrawnCount);

public sealed record ProviderOfflineServiceEnforcementMember(
    Brontide.Reference.Experimental.ComponentManagement.OccurrenceId Occurrence,
    string RetirementCode,
    bool ProviderStopped);

public sealed record ProviderOfflineServiceEnforcementResult(
    string Code,
    string RefusedBy,
    ProviderTrustOfflineDecision? Decision,
    IReadOnlyList<ProviderOfflineServiceEnforcementMember> Members,
    int AdmittedCount,
    int StoppedCount);

/// <summary>
/// Applies one offline-policy decision to the exact supplied serving set. Availability withdrawal
/// stops service but deliberately retains staged artifacts for the separate restart policy.
/// </summary>
public static class ProviderOfflineServiceEnforcement
{
    public const int MaximumMembers = 64;

    public static async ValueTask<ProviderOfflineServiceEnforcementResult> RunAsync(
        ProviderTrustOfflinePolicy policy,
        DateTimeOffset now,
        DateTimeOffset? lastCurrent,
        string pollCode,
        string? lastAttemptCode,
        IReadOnlyList<ProviderServingActivation> activations,
        string retirementReason)
    {
        ArgumentNullException.ThrowIfNull(policy);
        ArgumentNullException.ThrowIfNull(activations);
        ArgumentException.ThrowIfNullOrWhiteSpace(retirementReason);

        if (activations.Count > MaximumMembers
            || activations.Any(activation => activation is null || !activation.IsServing)
            || activations.Select(activation => activation.Occurrence).Distinct().Count() != activations.Count)
        {
            return new("offline-enforcement-invalid", "preflight", null,
                Array.Empty<ProviderOfflineServiceEnforcementMember>(), 0, 0);
        }

        var decision = policy.Evaluate(now, lastCurrent, pollCode, lastAttemptCode, activations.Count);
        if (decision.Code == "offline-existing-service")
            return new("offline-enforcement-continuing", "none", decision,
                Array.Empty<ProviderOfflineServiceEnforcementMember>(), activations.Count, 0);
        if (decision.Code == "offline-idle")
            return new("offline-enforcement-idle", "none", decision,
                Array.Empty<ProviderOfflineServiceEnforcementMember>(), 0, 0);

        var members = new List<ProviderOfflineServiceEnforcementMember>(activations.Count);
        foreach (var activation in activations.OrderBy(
                     activation => activation.Occurrence.Value, StringComparer.Ordinal))
        {
            var retirementCode = "retirement-not-attempted";
            if (activation.Lifecycle?.Member is { IsReleased: true } member)
            {
                try
                {
                    await member.RetireAsync(retirementReason, CancellationToken.None).ConfigureAwait(false);
                    retirementCode = "retired";
                }
                catch (Exception)
                {
                    retirementCode = "retirement-failed";
                }
            }

            var providerStopped = activation.Chain.Provider?.HasExited != false;
            if (activation.Chain.Provider is { HasExited: false } provider)
            {
                try
                {
                    await provider.DisposeAsync().ConfigureAwait(false);
                }
                catch (Exception)
                {
                    // Observe the concrete process after disposal: an exception does not prove that
                    // termination failed, and a successful return does not replace that observation.
                }
                providerStopped = provider.HasExited;
            }
            members.Add(new(activation.Occurrence, retirementCode, providerStopped));
        }

        var stopped = members.Count(member => member.ProviderStopped);
        var code = stopped != members.Count
            ? "offline-enforcement-incomplete"
            : members.Any(member => member.RetirementCode != "retired")
                ? "offline-enforcement-cleanup-incomplete"
                : "offline-enforcement-stopped";
        return new(code, "none", decision, members, activations.Count, stopped);
    }
}

/// <summary>
/// Applies one deterministic, bounded host-owned trust sweep. Invocation cadence remains external.
/// </summary>
public static class ProviderServingTrustSweep
{
    public const int MaximumMembers = 64;

    public static async ValueTask<ProviderServingTrustSweepResult> RunAsync(
        DurableProviderPublisherTrustPolicyRegistry registry,
        ContentAddressedProviderStore store,
        IReadOnlyList<ProviderServingActivation> activations,
        string retirementReason,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(activations);
        ArgumentException.ThrowIfNullOrWhiteSpace(retirementReason);

        if (activations.Count is < 1 or > MaximumMembers
            || activations.Any(activation => activation is null || !activation.IsServing)
            || activations.Select(activation => activation.Occurrence).Distinct().Count() != activations.Count)
        {
            return new(
                "serving-trust-sweep-invalid", "preflight",
                Array.Empty<ProviderServingTrustSweepMember>(), 0, 0);
        }

        ArgumentNullException.ThrowIfNull(registry);
        ArgumentNullException.ThrowIfNull(store);

        var ordered = activations
            .OrderBy(activation => activation.Occurrence.Value, StringComparer.Ordinal)
            .ToArray();
        var members = new List<ProviderServingTrustSweepMember>(ordered.Length);

        foreach (var activation in ordered)
        {
            var result = await ProviderServingTrustRevalidation.RevalidateAsync(
                registry, store, activation, retirementReason, removeStagedSet: false, cancellationToken)
                .ConfigureAwait(false);
            members.Add(new(activation.Occurrence, result));
        }

        foreach (var stagedGroup in ordered
                     .Where(activation => activation.Chain.StagedIdentity is not null)
                     .GroupBy(activation => activation.Chain.StagedIdentity!.Value))
        {
            var occurrences = stagedGroup.Select(activation => activation.Occurrence).ToHashSet();
            if (members.Any(member => occurrences.Contains(member.Occurrence)
                                      && (member.Result.Continued || !member.Result.Revalidated)))
            {
                continue;
            }

            var removal = store.Remove(stagedGroup.Key);
            if (!removal.Removed && removal.Code != "artifact-set-not-staged")
            {
                for (var index = 0; index < members.Count; index++)
                {
                    if (occurrences.Contains(members[index].Occurrence)
                        && !members[index].Result.Continued)
                    {
                        members[index] = members[index] with
                        {
                            Result = members[index].Result with
                            {
                                RetirementCode = $"{members[index].Result.RetirementCode};{removal.Code}",
                            },
                        };
                    }
                }
            }
        }

        var continued = members.Count(member => member.Result.Continued);
        var code = members.Any(member => !member.Result.Revalidated)
            ? "serving-trust-sweep-incomplete"
            : members.Any(member => !member.Result.Continued && member.Result.RetirementCode != "retired")
                ? "serving-trust-sweep-cleanup-incomplete"
                : continued == members.Count
                    ? "serving-trust-sweep-current"
                    : "serving-trust-sweep-withdrawn";

        return new(code, "none", members, continued, members.Count - continued);
    }
}
